src/core/cm/imagemanager/imagemanager.cpp

```diff
@@ -952,4 +1065,402 @@ Error ImageManager::CleanupOrphanedBlobs(
 }
 
+Error ImageManager::RemoveDifferentVersions(const Array<UpdateItemInfo>& itemsInfo, const Array<ItemInfo>& storedItems)
+{
+    LOG_DBG() << "Remove different versions";
+
+    for (const auto& storedItem : storedItems) {
+        auto requestedIt
+            = itemsInfo.FindIf([&storedItem](const auto& item) { return item.mItemID == storedItem.mItemID; });
+
+        if (requestedIt != itemsInfo.end()) {
+            if (requestedIt->mVersion != storedItem.mVersion) {
+                LOG_DBG() << "Wiping different version" << Log::Field("itemID", storedItem.mItemID)
+                          << Log::Field("storedVersion", storedItem.mVersion)
+                          << Log::Field("requestedVersion", requestedIt->mVersion);
+
+                if (auto err = mStorage->RemoveItem(storedItem.mItemID, storedItem.mVersion); !err.IsNone()) {
+                    LOG_ERR() << "Failed to remove item from storage" << Log::Field("itemID", storedItem.mItemID)
+                              << Log::Field("version", storedItem.mVersion) << Log::Field(err);
+                }
+
+                if (!storedItem.mIndexDigest.IsEmpty()) {
+                    auto filePath = fs::JoinPath(mBlobsInstallPath, storedItem.mIndexDigest);
+                    LOG_DBG() << "Wipe blob" << Log::Field("path", filePath);
+
+                    if (auto err = fs::RemoveAll(filePath); !err.IsNone()) {
+                        LOG_ERR() << "Failed to wipe blob" << Log::Field("path", filePath) << Log::Field(err);
+                    }
+                }
+            }
+        }
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ImageManager::VerifyBlobsIntegrity(const Array<UpdateItemInfo>& itemsInfo, Array<UpdateItemStatus>& statuses,
+    const Array<ItemInfo>& storedItems, Array<StaticString<oci::cDigestLen>>& requestedBlobs)
+{
+    LOG_DBG() << "Verify blobs integrity";
+
+    for (size_t i = 0; i < itemsInfo.Size(); i++) {
+        const auto& itemInfo = itemsInfo[i];
+
+        auto storedIt = storedItems.FindIf([&itemInfo](const auto& stored) {
+            return stored.mItemID == itemInfo.mItemID && stored.mVersion == itemInfo.mVersion;
+        });
+
+        if (storedIt == storedItems.end()) {
+            LOG_WRN() << "Item not found in storage" << Log::Field("itemID", itemInfo.mItemID)
+                      << Log::Field("version", itemInfo.mVersion);
+            statuses[i].mState = ItemStateEnum::eFailed;
+            statuses[i].mError = ErrorEnum::eNotFound;
+
+            continue;
+        }
+
+        Error integrityError = ErrorEnum::eNone;
+
+        auto removeOnFail = DeferRelease(storedIt, [this, &statuses, &integrityError, i](const ItemInfo* item) {
+            if (!integrityError.IsNone()) {
+                statuses[i].mState = ItemStateEnum::eFailed;
+                statuses[i].mError = integrityError;
+
+                if (auto err = mStorage->RemoveItem(item->mItemID, item->mVersion); !err.IsNone()) {
```

Comment:
Probably we need to set it to failed state instead of remove. Let's discuss.

Reply:
discussed to leave as is